newIDE/app/src/MainFrame/index.js

```diff
@@ -196,76 +201,125 @@ class MainFrame extends React.Component<Props, State> {
     gdjsDevelopmentWatcherEnabled: false,
     isPreviewFirstSceneOverriden: false,
     previewFirstSceneName: '',
-  };
-  toolbar = null;
-  _resourceSourceDialogs = {};
-  _previewLauncher: ?PreviewLauncherInterface = null;
-
-  componentWillMount() {
-    if (!this.props.integratedEditor) this.openStartPage();
-  }
-
-  componentDidMount() {
-    GD_STARTUP_TIMES.push(['MainFrameComponentDidMount', performance.now()]);
-
-    const { initialFileMetadataToOpen } = this.props;
-
-    this._loadExtensions()
-      .catch(() => {
-        /* Ignore errors */
-      })
-      .then(() => {
-        // Enable the GDJS development watcher *after* the extensions are loaded,
-        // to avoid the watcher interfering with the extension loading (by updating GDJS,
-        // which could lead in the extension loading failing for some extensions as file
-        // are removed/copied).
-        this.setState({
-          gdjsDevelopmentWatcherEnabled: true,
-        });
-      });
-    if (initialFileMetadataToOpen) {
-      this._openInitialFileMetadata(/* isAfterUserInteraction= */ false);
-    } else if (this.props.introDialog && !Window.isDev())
-      this._openIntroDialog(true);
-
-    GD_STARTUP_TIMES.push([
-      'MainFrameComponentDidMountFinished',
-      performance.now(),
-    ]);
-    console.info('Startup times:', getStartupTimesSummary());
-  }
+  });
+  const toolbar = React.useRef(null);
+  const _resourceSourceDialogs = React.useRef({});
+  const _previewLauncher = React.useRef((null: ?PreviewLauncherInterface));
+  //const { addRecentFile } = React.useContext(PreferencesContext);
+  const forceUpdate = useForceUpdate();
+
+  React.useEffect(() => {
+    // This is just for testing, to check if we're getting the right state
+    // and gives us an idea about the number of re-renders.
+    console.log(state);
+  });
+
+  const _openIntroDialog = React.useCallback(
+    (open: boolean = true) => {
+      setState(state => ({ ...state, introDialogOpen: open }));
+    },
+    [setState]
+  );
+
+  const { integratedEditor, initialFileMetadataToOpen, introDialog } = props;
+  React.useEffect(
+    () => {
+      (() => {
+        if (!integratedEditor) openStartPage();
+        GD_STARTUP_TIMES.push([
+          'MainFrameComponentDidMount',
+          performance.now(),
+        ]);
+        _loadExtensions()
+          .then(() =>
+            // Enable the GDJS development watcher *after* the extensions are loaded,
+            // to avoid the watcher interfering with the extension loading (by updating GDJS,
+            // which could lead in the extension loading failing for some extensions as file
+            // are removed/copied).
+            setState(state => ({
+              ...state,
+              gdjsDevelopmentWatcherEnabled: true,
+            }))
+          )
+          .then(state => {
```

Comment:
this argument state is probably undefined, because the previous then is not returning anything. Let's remove it.

Reply:
I missed than the previous function did not need a return. This is fine then!